chain/index/events.go

```diff
@@ -0,0 +1,548 @@
+			}
+
+			// Insert event into events table
+			eventResult, err := tx.Stmt(si.stmts.insertEventStmt).ExecContext(ctx, messageID, eventCount, uint64(event.Emitter), robustAddrbytes, 0)
+			if err != nil {
+				return xerrors.Errorf("failed to insert event: %w", err)
+			}
+
+			// Get the event_id of the inserted event
+			eventID, err := eventResult.LastInsertId()
+			if err != nil {
+				return xerrors.Errorf("failed to get last insert id for event: %w", err)
+			}
+
+			// Insert event entries
+			for _, entry := range event.Entries {
+				_, err := tx.Stmt(si.stmts.insertEventEntryStmt).ExecContext(ctx,
+					eventID,
+					isIndexedValue(entry.Flags),
+					[]byte{entry.Flags},
+					entry.Key,
+					entry.Codec,
+					entry.Value,
+				)
+				if err != nil {
+					return xerrors.Errorf("failed to insert event entry: %w", err)
+				}
+			}
+			eventCount++
+		}
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) loadExecutedMessages(ctx context.Context, msgTs, rctTs *types.TipSet) ([]executedMessage, error) {
+	msgs, err := si.cs.MessagesForTipset(ctx, msgTs)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get messages for tipset: %w", err)
+	}
+
+	st := si.cs.ActorStore(ctx)
+
+	receiptsArr, err := blockadt.AsArray(st, rctTs.Blocks()[0].ParentMessageReceipts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load message receipts array: %w", err)
+	}
+
+	if uint64(len(msgs)) != receiptsArr.Length() {
+		return nil, xerrors.Errorf("mismatching message and receipt counts (%d msgs, %d rcts)", len(msgs), receiptsArr.Length())
+	}
+
+	ems := make([]executedMessage, len(msgs))
+
+	for i := 0; i < len(msgs); i++ {
+		ems[i].msg = msgs[i]
+
+		var rct types.MessageReceipt
+		found, err := receiptsArr.Get(uint64(i), &rct)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to load receipt %d: %w", i, err)
+		}
+		if !found {
+			return nil, xerrors.Errorf("receipt %d not found", i)
+		}
+		ems[i].rct = rct
+
+		// no events in the receipt
+		if rct.EventsRoot == nil {
+			continue
+		}
+
+		eventsArr, err := amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+		if err != nil {
+			if si.recomputeTipSetStateFunc == nil {
+				return nil, xerrors.Errorf("failed to load events amt for message %s: %w", ems[i].msg.Cid(), err)
+			}
+			log.Warnf("failed to load events amt for message %s: %s; recomputing tipset state to regenerate events", ems[i].msg.Cid(), err)
+
+			if err := si.recomputeTipSetStateFunc(ctx, msgTs); err != nil {
+				return nil, xerrors.Errorf("failed to recompute missing events; failed to recompute tipset state: %w", err)
+			}
+
+			eventsArr, err = amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+			if err != nil {
+				return nil, xerrors.Errorf("failed to load events amt for message %s: %w", ems[i].msg.Cid(), err)
+			}
+
+			log.Infof("successfully recomputed tipset state and loaded events amt for message %s", ems[i].msg.Cid())
```

Comment:
What are the cases where we might get this message? We're going to get spammed with this for a backfill from a snapshot aren't we because they won't come with the events, right? This would be interesting to see, but it could get very noisy for each message where we run into cases that need this.

Reply:
@rvagg 

For snapshot hydration -> this code path dosen't get hit all. See the implementation of `PopulateFromSnapshot` which explicitly avoids indexing events.

**_The case where this can happen is:_**
- User had `storeEvents` OFF for some time for whatever reason (for eg: they had their `ETHRPCAPI` disabled for some reason) and have now turned it on -> User kicks off a backfilling for epochs during the duration where `storeEvents` was disabled -> we don't have events in the AMT -> need to do a re-computation.

Let's keep this long line around and then turn it OFF if we see a complaint.

Reply:
yeah ok, reasonable

we should look at the PopulateFromSnapshot & events though, glif rpc nodes will probably want that, for instance, maybe a setting that lets you do this but you have to accept that it'll take time -- we should find out I guess

Reply:
Let me ask RPC providers.

Reply:
Asked  Protofire for feedback at https://filecoinproject.slack.com/archives/C03TM2QDTHN/p1728977815186839.

Reply:
Glif Nodes: We receive a lot of requests for logs and events at our public endpoints. Multiple nodes backs each endpoint, and we use sticky sessions for data consistency. It would be weird if most users got normal responses and some users got empty ones just because they were unlucky enough to be stuck to a node that was re-synced from the snapshot less than 16 hours ago, which is what will happen if we don't have events available at the node start.

Like, one thing we can do is not let node in the active set before it indexes back events, but it would increase the wait for the node for 30 minutes as you say, which is always not preferrable :slightly_smiling_face:

Reply:
@ArseniiPetrovich what do you do at the moment to sync a newly provisioned node? do you run `lotus-shed indexes backfill-*` on it? I think right now you won't get events for the 900 epochs in the snapshot you're restoring from if you're using `--import-snapshot` method during provisioning so are you currently adressing that?

Reply:
@rvagg I think the best path ahead here is for RPC providers to use  the `lotus-shed backfill` `ChainIndexer` command to backfill missing events for the range of epochs in the Snapshot if they so wish to do so.

This leaves the choice in the hand of RPC providers and ALSO dosen't penalise the snapshot import. The backfilling does recompute tipsets to generate events if event storage is enabled and so this will work fine.